### PR TITLE
Devise authentication implementation

### PR DESCRIPTION
**Pull request by bennacer860**

### CODE REVIEW

File: app/views/diagrams/index.html.erb [hunk 1]
```diff
@@ -1,19 +1,29 @@
...
+<br>
+<br>
 
+<% if current_user.present? %>
+  Your current role is: <%= current_user.role %>
+<% end %>
 
-<div class="card-columns">
-  <% @diagrams.shuffle.each do |diagram| %>
+<% @diagrams.each do |diagram| %>
```
Comment (bennacer860) [inline app/views/diagrams/index.html.erb:11]: Why did you remove the shuffle method? It is suppose to render a random order of the diagrams

File: app/views/diagrams/index.html.erb [hunk 2]
```diff
@@ -1,19 +1,29 @@
...
 
+        <% if diagram.user == current_user || current_user&.admin? %>
+          <%= link_to 'Edit', edit_diagram_path(diagram), :class => "btn btn-sm btn-outline-secondary" %>
+          <%= link_to 'Destroy', diagram, method: :delete, data: { confirm: 'Are you sure?' }, :class => "btn btn-sm btn-outline-danger" %></p>    
+        <% end %>
       <p class="card-text"><small class="text-muted">Last updated <%= time_ago_in_words(diagram.created_at) %> ago</small></p>
-    </div>
+      <p class="card-text"><small class="text-muted">created by <%= diagram.user.email if diagram.user %></small></p>
+
+      </div>
```
Comment (bennacer860) [inline app/views/diagrams/index.html.erb:26]: Please fix identation

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -26,70 +25,15 @@
...
-
   <!--Main layout-->
+  
   <div class="container">
-    <%= yield %>
+  <%= render 'layouts/navbar' %>
   </div>
 
-
+    <%= yield %>
```
Comment (bennacer860) [inline app/views/layouts/application.html.erb:34]: Only `<%= yield %>` should be in the container block, please remove the `  <%= render 'layouts/navbar' %>`

File: db/migrate/20230503180255_add_role_to_users.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddRoleToUsers < ActiveRecord::Migration[6.1]
+  def change
+    add_column :users, :role, :integer, default: 0
```
Comment (bennacer860) [inline db/migrate/20230503180255_add_role_to_users.rb:3]: This will not collect any useful information for now, but we can deal with it in another PR

File: app/controllers/diagrams_controller.rb [hunk 5]
```diff
@@ -49,14 +51,20 @@ def update
...
   def destroy
-    @diagram.destroy
-
     respond_to do |format|
       format.html { redirect_to diagrams_url, notice: "Diagram was successfully destroyed." }
       format.json { head :no_content }
     end
   end
 
+  def require_permission
```
Comment (bennacer860) [inline app/controllers/diagrams_controller.rb:60]: Please delete this method since it is define in line 79

File: app/controllers/diagrams_controller.rb [hunk 6]
```diff
@@ -49,14 +51,20 @@ def update
 
   # DELETE /diagrams/1 or /diagrams/1.json
   def destroy
-    @diagram.destroy
```
Comment (bennacer860) [inline app/controllers/diagrams_controller.rb:52]: Please bring back the destroy method, otherwise the `destroy` controller action will be a no-operation

**Review by bennacer860 — COMMENTED** [review]
Overall it is looking good, only minor changes needs to be implemented. 

There is no commit for the `Gemfile` and the `Gemfile.lock`, did you forget them?

File: app/models/diagram.rb [hunk 7]
```diff
@@ -10,4 +10,11 @@ def format_source_code_to_html
       "#{'&nbsp;'*leading_spaces_size}#{line}"
     end.join('<br/>')
   end
+
+
+  belongs_to :user
```
[the remark this answers is not in the document]
Comment (AmarBennacer) [reply app/models/diagram.rb:15]: done-move to line 2

File: app/models/user.rb [hunk 8]
```diff
@@ -3,4 +3,14 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
+
+ enum role: [:user, :admin]
```
[the remark this answers is not in the document]
Comment (AmarBennacer) [reply app/models/user.rb:7]: fixed

File: app/views/diagrams/index.html.erb (hunk 1 above)
Comment (AmarBennacer) [reply app/views/diagrams/index.html.erb:11]: added shuffle function

File: app/views/diagrams/index.html.erb (hunk 2 above)
Comment (AmarBennacer) [reply app/views/diagrams/index.html.erb:26]: fixed

File: db/migrate/20230503180255_add_role_to_users.rb (hunk 4 above)
Comment (AmarBennacer) [reply db/migrate/20230503180255_add_role_to_users.rb:3]: ok

File: app/controllers/diagrams_controller.rb (hunk 5 above)
Comment (AmarBennacer) [reply app/controllers/diagrams_controller.rb:60]: removed

File: app/controllers/diagrams_controller.rb (hunk 6 above)
Comment (AmarBennacer) [reply app/controllers/diagrams_controller.rb:52]: addde back-

Comment (AmarBennacer) [reply app/controllers/diagrams_controller.rb:52]: done

**Comment by AmarBennacer** [comment]
Modifeid code according to comments. 
Can you confirm gemfile and gemfile.lock are okay as I see them tracked on my local terminal
